src/iop/flip.c

```diff
@@ -518,6 +518,14 @@ void reload_defaults(dt_iop_module_t *self)
   }
 }
 
+static void _crop_callback(const int mode)
```

Comment:
For now i have just changed the `_crop_callback` parameters as suggested but left out the "compression".

Reply:
My suggestion here can be simplified, and made more robust, if the order the history items are created in is reversed (i.e. first a crop adjustment if needed and then the orientation). Then we can just test if the latest item is an orientation and "remove" it, by decreasing history_end. _If_ there is an existing crop and _if_ it needs adjusting, then the `dt_dev_add_history_item` in `flip_callback` will automatically merge them. A new orientation item will always be needed.

A problem with the approach I described above (having an orientation+crop) is that it is possible that after changing the orientation (and having a crop automatically generated) the user could then reset the crop, which would just amend the automatically generated crop history item. Then another orientation change would first remove the crop item, but then find, since it was at default, that it is not necessary to create a new adjusted crop. But now the crop reset has disappeared from history. This problem doesn't exist if the two history items are reversed.

Reply:
Yes, this might get pretty tricky if after flip&crop the crop is changed.

What about history items not to be selectable in the history list (maybe faded out) or even hidden history items?

Reply:
Yes, this might get pretty tricky if after flip&crop the crop is changed.

What about history items not to be selectable in the history list (maybe faded out) or even hidden history items?

Reply:
> Yes, this might get pretty tricky if after flip&crop the crop is changed.

My point is that this is _not_ a problem at all if the order of the history items is reversed:
```
  if(cropper && dev->cropping.flip_callback)
    dev->cropping.flip_callback(cropper, mode);
  dt_dev_add_history_item(darktable.develop, self, TRUE);
```
Then the "orientation" item is the last one in history and if you then adjust the crop, it will just create a new item (instead of merging with an existing crop item). _Then_ the adjusted crop will be the last item and a new "orientation" change will test and see that the last item is _not_ orientation and not remove/replace it and two new items will be created.

> What about history items not to be selectable in the history list (maybe faded out) or even hidden history items?

What? Are you serious?

Reply:
Not fully serious. Just an outspoken idea about how could we implement combined-module-action.

Reply:
Reverse Order of applied history might be good anyway.

Reply:
Now tested and
```
  dt_pthread_mutex_lock(&dev->history_mutex);
  GList *last = g_list_nth(dev->history, dev->history_end - 1);
  if(last && ((dt_dev_history_item_t *)last->data)->module == self)
    dev->history_end--; // don't create duplicate orientation+crop pairs
  dt_pthread_mutex_unlock(&dev->history_mutex);

  if(cropper && dev->cropping.flip_callback)
    dev->cropping.flip_callback(cropper, mode);
  dt_dev_add_history_item(dev, self, TRUE);
```
works fine, _except_ that it breaks _undo_. Enclosing in `dt_undo_start_group` and `dt_undo_end_group` doesn't seem to solve that problem and I don't have further ideas at the moment...

Reply:
Not sure atm what the better order of crop<->flip in history stack is. Also it seems better to me right now to not-compress history so undo/redo works just fine and as before.